### PR TITLE
layer-wise quant

### PR DESCRIPTION
**Pull request by n1ck-guo**
## Type of Change

new feature

## Description

layer by layer quantization

## Expected Behavior & Potential Risk

No

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

accelerate

### CODE REVIEW

**Comment by xin3he** [comment]
Some unittest is required for layer-wise quant

File: examples/pytorch/nlp/huggingface_models/language-modeling/quantization/layer-wise/run_clm.py [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+        tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path, **tokenizer_kwargs)
+    else:
+        raise ValueError(
+            "You are instantiating a new tokenizer from scratch. This is not supported by this script."
+            "You can do it from another script, save it, and load it from here, using --tokenizer_name."
+        )
+
+    if model_args.model_name_or_path:
+        if model_args.tune:
+            from neural_compressor.compression.layer_wise_quant import load_shell
```
Comment (wenhuach21) [inline examples/pytorch/nlp/huggingface_models/language-modeling/quantization/layer-wise/run_clm.py:392]: better move this function to torch_utis/torch_utils/layer_wise_quant

File: neural_compressor/adaptor/torch_utils/layer_wise_quant/modified_pickle.py [hunk 2]
```diff
@@ -0,0 +1,1802 @@
+"""Create portable serialized representations of Python objects.
+
+See module copyreg for a mechanism for registering custom picklers.
+See module pickletools source for extensive comments.
+
+Classes:
+
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/layer_wise_quant/modified_pickle.py:7]: please be careful of the license. I assume most of the code is borrowed from other lib.

File: neural_compressor/adaptor/torch_utils/layer_wise_quant/quantize.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+from copy import deepcopy
+from tqdm import tqdm
+
+from .utils import torch
+from torch.quantization import prepare, convert
+from accelerate.utils import set_module_tensor_to_device
+from .utils import _get_path, get_named_children, update_module, load_tensor_from_shard, load_tensor
+
+
+TMP_DIR = f'./layer_wise_quant_tmp_dir_{time.time()}'
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/layer_wise_quant/quantize.py:14]: inc should have a workdir, you could use the same one in the future.

File: neural_compressor/adaptor/torch_utils/layer_wise_quant/torch_load.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        def find_class(self, mod_name, name):
+            if type(name) is str and 'Storage' in name:
+                try:
+                    return StorageType(name)
+                except KeyError:
+                    pass
+            mod_name = load_module_mapping.get(mod_name, mod_name)
+            return super().find_class(mod_name, name)
+
+        def persistent_load(self, saved_id):
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/layer_wise_quant/torch_load.py:93]: license

**Review by wenhuach21 — APPROVED** [review]

File: examples/pytorch/nlp/huggingface_models/language-modeling/quantization/layer-wise/run_clm.py (hunk 1 above)
Comment (n1ck-guo) [reply examples/pytorch/nlp/huggingface_models/language-modeling/quantization/layer-wise/run_clm.py]: This function is in torch_utils/layer_wise_quant, I forget to remove the example. New example will be added in ITREX.

File: neural_compressor/adaptor/torch_utils/layer_wise_quant/modified_pickle.py [hunk 5]
```diff
@@ -0,0 +1,1819 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2021 Intel Corporation
```
Comment (chensuyue) [inline neural_compressor/adaptor/torch_utils/layer_wise_quant/modified_pickle.py:4]: # Copyright (c) 2023 Intel Corporation

File: neural_compressor/adaptor/torch_utils/layer_wise_quant/__init__.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2021 Intel Corporation
```
Comment (chensuyue) [inline neural_compressor/adaptor/torch_utils/layer_wise_quant/__init__.py:4]: # Copyright (c) 2023 Intel Corporation

File: neural_compressor/adaptor/torch_utils/layer_wise_quant/quantize.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+def mk_tmp_dir():
+    os.makedirs(TMP_DIR, exist_ok=True)
+
+
+def del_tmp_dir():
+    shutil.rmtree(TMP_DIR)
+
+
+class QDQLayer(torch.nn.Module):
```
Comment (xin3he) [inline neural_compressor/adaptor/torch_utils/layer_wise_quant/quantize.py:43]: How about move this QDQLayer to torch_utils/model_wrapper.py

**Review by xin3he — APPROVED** [review]

**Pull request merged**